### PR TITLE
Add support for `infer` in conditionals operating on function parameters

### PR DESCRIPTION
**Pull request by ninjack-dev**
Adds support for types which use `infer` in the function parameters, specifically where `infer` is used on the only parameter; in short, adds support for `Parameters<T>`.

**Notes**:
- Please see [this line in FunctionNodeParser](https://github.com/ninjack-dev/ts-json-schema-generator/blob/1124d59eefe1995847f72fc45620bacfb14ad3c2/src/NodeParser/FunctionNodeParser.ts#L60); to cover the base `InferType` case when handling function parameter types, I intersect the resulting `InferType` with `ObjectType` to pass the requirements for `FunctionType`. This is a bit of a hack, but it works well enough; curious how I should go about handling this, if at all.
- Added tests are minimal, but I imagine `Parameters<T>` makes up 90% of the use cases for this functionality (or at least it does for mine); if edge cases/other uses are discovered later I'll be happy to revisit this and add more tests.

### CODE REVIEW

File: src/NodeParser/FunctionNodeParser.ts [hunk 1]
```diff
@@ -53,6 +54,12 @@ export function getNamedArguments(
         return undefined;
     }
 
+    // Special case for when function signature is (...args: infer T)
+    if (node.parameters.length === 1) {
+        const parameterType = childNodeParser.createType(node.parameters[0], context);
+        if (parameterType instanceof InferType) return parameterType as typeof parameterType & ObjectType;
```
Comment (arthurfiorette) [inline src/NodeParser/FunctionNodeParser.ts:60]: Is this cast safe? how can parameterType be InferType and ObjectType at the same time?

File: test/valid-data/type-conditional-infer-function/schema.json [hunk 2]
```diff
@@ -0,0 +1,22 @@
+{
+  "$ref": "#/definitions/FuncParams",
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "definitions": {
+    "FuncParams": {
```
Comment (arthurfiorette) [inline test/valid-data/type-conditional-infer-function/schema.json:5]: In this case, shouldn't FuncParams be a `[string, number]` tuple?

**Review by arthurfiorette — CHANGES_REQUESTED** [review]

File: test/valid-data/type-conditional-infer-function/schema.json (hunk 2 above)
Comment (ninjack-dev) [reply test/valid-data/type-conditional-infer-function/schema.json:5]: I suppose that would be in line with the expected result of a tuple type, looking at the other tests.

Intuitively, a method signature in something like OpenAPI would be made up of base primitive/object schemas instead of labeled tuple values, which is why I initially targeted this; is the recommendation here to just use a custom formatter?

Comment (arthurfiorette) [reply test/valid-data/type-conditional-infer-function/schema.json:5]: test/valid-data/type-aliases-tuple-optional-items/schema.json is one example of tuple generated types. But I'm not sure if its the best way to represent down parameters since order matters for them.

@domoritz any considerations here?

File: src/NodeParser/FunctionNodeParser.ts (hunk 1 above)
Comment (ninjack-dev) [reply src/NodeParser/FunctionNodeParser.ts:60]: I'm not 100% sure, but I think it's safe for now. I did a cursory reference check for where `FunctionType`s are used and couldn't find a scenario where the intersection is invalid. Also, TSC didn't report anything and neither did ESLint.

I can't think of a test case that would cause this to break, and all current tests pass with it given the scenarios it would actually happen in (i.e. `Parameters<T>`); I tried other arbitrary tests using some form of `(...args: infer P) => any`, but I couldn't find a way to force usage of the `ObjectType` properties that aren't common to `InferType`. Happy to contribute a fix if this causes issues for a new test down the line.

File: test/valid-data/type-conditional-infer-function/schema.json (hunk 2 above)
Comment (domoritz) [reply test/valid-data/type-conditional-infer-function/schema.json:5]: There is not standard so I don't have strong opinions as long as we are consistent.

Comment (arthurfiorette) [reply test/valid-data/type-conditional-infer-function/schema.json:5]: Im Ok with Parameters<> itself returning an object since naming and ordering kinda matters as long its just for Parameters<>. Maybe instead of this all we can create a custom nodeparser for Parameters<> to isolate this logic just in there.

**Review by arthurfiorette — CHANGES_REQUESTED** [review]

File: test/valid-data/type-conditional-infer-function/schema.json (hunk 2 above)
Comment (ninjack-dev) [reply test/valid-data/type-conditional-infer-function/schema.json:5]: A `Parameters<T>`-specific parser wouldn't be appropriate IMO; there are other scenarios in which `infer P` could be used. Any conditional type which would e.g. compare function signatures would also use it (this example is admittedly poor; see collapsed details for notes):
```ts
type HasReturnType<T, U> = T extends ((...args: infer P) => U) ? ((...args: P) => U) : never;
export type Check = HasReturnType<((a: number) => void), void>
```

The ultimate problem is that TypeScript seems to internally represents the parameter type as a tuple, while the `ts-json-schema-generator` parses it as an object (which I certainly prefer from an IPC standpoint). To allow the generation of named tuples with `infer P` would require the function parser to be changed outright to reflect what Typescript uses internally; in my opinion, it's ultimately a source-of-truth problem, and I'm inclined to stick with the current implementation for simplicity and usability.

<details>

`Check` produces some wonky output:
  ```json
  
  {
    "$ref": "#/definitions/Check",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "definitions": {
      "Check": {
        "$comment": "(...args: P) => U",
        "properties": {
          "namedArgs": {
            "additionalProperties": false,
            "properties": {
              "args": {
                "additionalProperties": false,
                "properties": {
                  "a": {
                    "type": "number"
                  }
                },
                "required": [
                  "a"
                ],
                "type": "object"
              }
            },
            "required": [
              "args"
            ],
            "type": "object"
          }
        },
        "type": "object"
      }
    }
  }
  ```
In fairness, I don't know why you would use `Check` from this example to generate a schema. Hard to say if it's worth fixing; I'll try to find a more appropriate scenario to justify it.
</details>

Comment (arthurfiorette) [reply test/valid-data/type-conditional-infer-function/schema.json:5]: My concern was about infer in general, but this PR only affects infer inside function parameters, and in this case I agree that it should generate an object.

**Review by arthurfiorette — APPROVED** [review]
